plugins/glitch/main.py

```diff
@@ -0,0 +1,53 @@
+import asyncio
+import functools
+import yaml
+
+from .selectors import SELECTOR_MAP as selector_map
+from .actions import ACTION_MAP as action_map
+
+DEFAULT_OUTPUT = '/tmp/last_glitch_plan.yaml'
+
+
+async def glitch(context, rule, action):
+    '''
+
+
+    '''
+    rule.log.info("Starting glitch")
+
+    glitch_output = []
+    output_filename = DEFAULT_OUTPUT
+
+    glitch_plan = action.args.get('glitch_plan') or generate_plan(
+        context.model,
+        num=action.args.get('glitch_num', 5),
+        action_map=action_map
+    )
+    rule.log.info("Writing glitch plan to {}".format(output_file))
+    with open(output_filename, 'w') as output_file:
+        output_file.write(yaml.dump(glitch_plan))
+
+    # Execute glitch plan. We perform destructive operations here!
+    for action in glitch_plan:
+        selectors = action.pop('selectors')
+        actionf = action_map[action.pop('action')]
+
+        # Find a set of units to act upon
+        units = None
+        for selector in selectors:
+            selectf = selector_map[selector.pop['selector']]
+            units = selectf(context.model, **selector)
```

Comment:
bcsaller: this makes a lot of sense. In the latest iteration of the code, I just added hooks for looping over a set, but everything is nicely encapsulated so that we can add the other hooks that you suggested around it, as we need them.